### PR TITLE
fix adamw apply gradient

### PR DESCRIPTION
**Pull request by wangxicoding**
### PR types

Bug fixes
### PR changes

APIs
### Describe

1. Fix Adamw weight_decay not applied when open AMP. see https://github.com/PaddlePaddle/Paddle/issues/29794
2. Fix Adamw weight_decay LR is fixed when using LRSchedule.
3. Fix Adam step() not add imperative_base.no_grad, which gradients may be generated in the optimizer.
4. Optimization calculation, the calculation formula of `param = param - param * lr * coeff` is optimized as follows
     `param = param * (1.0 - lr * coeff)`

### CODE REVIEW

File: python/paddle/optimizer/adamw.py [hunk 1]
```diff
@@ -139,96 +140,55 @@ def __init__(self,
...
+
+        if self._apply_decay_param_fun is not None \
+                and not self._apply_decay_param_fun(param.name):
+            return
+
+        if isinstance(self._coeff, float):
+            assert param.dtype == paddle.fluid.core.VarDesc.VarType.FP32, \
+                "the type of coeff(float) and parameter(%s) is not consistent."%(param.dtype)
+        else:
+            assert self._coeff.dtype == param.dtype, \
```
Comment (guoshengCS) [inline python/paddle/optimizer/adamw.py]: 感觉用户使用double有点麻烦，能否float时就不要求param.dtype呢，下面的`decay_coeff = 1.0 - self._coeff * learning_rate`一定会受影响而不是用learning_rate的dtype是吗

Comment (wangxicoding) [reply python/paddle/optimizer/adamw.py]: 是指把这整个判断逻辑都干掉吗

Comment (wangxicoding) [reply python/paddle/optimizer/adamw.py]: Done.
1.0 - self._coeff * learning_rate，返回的tensor类型是self._coeff的类型，如果self._coeff 和 learning_rate类型不一致，会自动加cast转换类型。
把判断逻辑去掉则支持任何类型，改成1.0 - learning_rate * self._coeff使用learning_rate的类型。

**Review by guoshengCS — DISMISSED** [review]
LGTM

**Review by swtkiwi — DISMISSED** [review]
LGTM

**Review by guoshengCS — APPROVED** [review]

**Review by swtkiwi — APPROVED** [review]
LGTM
